### PR TITLE
[SpannerToSourceDb] Improving UT coverage for SpannerToSourceDb template - DML generators

### PR DESCRIPTION
**Pull request by darshan-sj**

### CODE REVIEW

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java [hunk 1]
```diff
@@ -872,4 +872,189 @@ public void testSourcePkNull() {
...
+
+  @Test
+  public void testGetDeleteStatementCQL_NullTimestamp() {
+    Map<String, PreparedStatementValueObject<?>> pkValues =
+        Map.of("id", PreparedStatementValueObject.create("int", 1));
+    DMLGeneratorResponse response =
+        CassandraDMLGenerator.getDeleteStatementCQL("my_table", null, pkValues);
+
+    assertEquals(
+        "DELETE FROM \"my_table\" USING TIMESTAMP ? WHERE \"id\" = ?", response.getDmlStatement());
```
[the remark this answers is not in the document]
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java:905]: Fixed.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java [hunk 2]
```diff
@@ -872,4 +872,189 @@ public void testSourcePkNull() {
...
+  }
+
+  @Test
+  public void testGetColumnValues_MissingSpannerColumnMapping() throws Exception {
+    com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper mockSchemaMapper =
+        org.mockito.Mockito.mock(
+            com.google.cloud.teleport.v2.spanner.migrations.schema.ISchemaMapper.class);
```
[the remark this answers is not in the document]
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/MySQLDMLGeneratorTest.java [hunk 3]
```diff
@@ -1557,4 +1557,177 @@ public void testGeneratedPrimaryKeyDML() {
     assertEquals(0, countInSQL(sql, "FirstName"));
     assertEquals(0, countInSQL(sql, "SingerId"));
   }
+
+  @Test
+  public void testBitTypeDML() throws Exception {
+    com.google.cloud.teleport.v2.spanner.ddl.Column spannerCol =
+        org.mockito.Mockito.mock(com.google.cloud.teleport.v2.spanner.ddl.Column.class);
```
[the remark this answers is not in the document]
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/MySQLDMLGeneratorTest.java]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java [hunk 4]
```diff
@@ -445,4 +445,257 @@ public void testNonByteaDecodeBranch() {
...
+    String res = PostgreSQLDMLGenerator.getMappedColumnValue(spannerCol, sourceCol, json, "+00:00");
+    assertEquals("'{\"a\": 1}'", res);
+  }
+
+  @Test
+  public void testGetDMLStatement_SourceTableNotFound() {
+    PostgreSQLDMLGenerator generator = new PostgreSQLDMLGenerator();
+    com.google.cloud.teleport.v2.templates.models.DMLGeneratorRequest request =
+        org.mockito.Mockito.mock(
+            com.google.cloud.teleport.v2.templates.models.DMLGeneratorRequest.class);
```
[the remark this answers is not in the document]
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java [hunk 5]
```diff
@@ -872,4 +879,137 @@ public void testSourcePkNull() {
                     .setSourceSchema(sourceSchema)
                     .build()));
   }
+
+  @Test
+  public void testGetDeleteStatementCQL_NullTimestamp() {
```
Comment (shreyakhajanchi) [inline v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java:887]: can we add a test where timestamp is not null?

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+ */
+package com.google.cloud.teleport.v2.templates.dbutils.dml;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThrows;
+
+import org.junit.Test;
+
+public class DMLGeneratorUtilsTest {
```
Comment (shreyakhajanchi) [inline v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java:34]: why are we not adding tests for `getColumnValues` and `getPkColumnValues` that is the majority code in DMLGeneratorUtils?

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java [hunk 7]
```diff
@@ -445,4 +445,210 @@ public void testNonByteaDecodeBranch() {
...
+    when(request.getSchemaMapper()).thenReturn(schemaMapper);
+    when(request.getSourceSchema()).thenReturn(sourceSchema);
+    when(request.getSpannerDdl()).thenReturn(ddl);
+    when(ddl.table("Singers")).thenReturn(spannerTable);
+
+    try {
+      when(schemaMapper.getSourceTableName("", "Singers")).thenReturn("Singers");
+    } catch (Exception e) {
+      // ignore
+    }
```
Comment (shreyakhajanchi) [inline v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: why do we need a try-catch block here?

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java [hunk 8]
```diff
@@ -445,4 +445,210 @@ public void testNonByteaDecodeBranch() {
...
+        postgreSQLDMLGenerator.getDMLStatement(
+            new DMLGeneratorRequest.Builder(
+                    modType, tableName, newValuesJson, keyValuesJson, "+00:00")
+                .setSchemaMapper(schemaMapper)
+                .setDdl(ddl)
+                .setSourceSchema(sourceSchema)
+                .build());
+    String sql = dmlGeneratorResponse.getDmlStatement();
+
+    assertTrue(sql.contains("NULL")); // Line 133!
```
Comment (shreyakhajanchi) [inline v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: I don't think we should mention line numbers in comments as it is subject to change in future, same comment for other places

Comment (shreyakhajanchi) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: Also can we test LastName=NULL ?

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java (hunk 5 above)
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/CassandraDMLGeneratorTest.java:887]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java (hunk 6 above)
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java:34]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java (hunk 8 above)
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: Done.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java (hunk 7 above)
Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/PostgreSQLDMLGeneratorTest.java]: Removed. And removed all such occurrences.

File: v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    when(schemaMapper.getSpannerColumnName("", "src_table", "col1")).thenReturn("spanner_col1");
+    when(sourceTable.name()).thenReturn("src_table");
+    when(spannerTable.column("spanner_col1")).thenReturn(spannerCol);
+    when(spannerCol.name()).thenReturn("spanner_col1");
+
+    JSONObject keyValuesJson = new JSONObject();
+    keyValuesJson.put("spanner_col1", "val1");
+
+    DMLGeneratorUtils.ColumnValueMapper mapper = (s, c, j, t) -> "mapped_val";
```
Comment (shreyakhajanchi) [inline v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java]: This might result in false positives, can we do something like:
```
DMLGeneratorUtils.ColumnValueMapper mapper =
    (spannerColDef, sourceColDef, valuesJson, timezoneOffset) -> {
      assertEquals(spannerCol, spannerColDef);
      assertEquals(sourceCol, sourceColDef);
      assertEquals(keyValuesJson, valuesJson);
      assertEquals("+00:00", timezoneOffset);
      return "mapped_val";
    };
```

Comment (darshan-sj) [reply v2/spanner-to-sourcedb/src/test/java/com/google/cloud/teleport/v2/templates/dbutils/dml/DMLGeneratorUtilsTest.java]: Done.

**Review by shreyakhajanchi — APPROVED** [review]
